pkg/governance-scripts/contracts/ProtocolFeeControllerMigrationV2.sol

```diff
@@ -0,0 +1,111 @@
+ *
+ * When all pools have been migrated, call `finalizeMigration` to disable further migration, update the address in the
+ * Vault, and renounce all permissions. While `migratePools` is permissionless, this call must be permissioned to
+ * prevent premature termination in case multiple transactions are required to migrate all the pools.
+ *
+ * Associated with `20250221-protocol-fee-controller-migration` (fork test only).
+ */
+contract ProtocolFeeControllerMigrationV2 is ProtocolFeeControllerMigration, SingletonAuthentication {
+    // Set after the global percentages have been transferred (on the first call to `migratePools`).
+    bool internal _globalPercentagesMigrated;
+
+    // ActionId for permission required in `migratePools`.
+    bytes32 internal _migrationRole;
+
+    /// @notice Cannot call the base contract migration; it is invalid for this migration.
+    error WrongMigrationVersion();
+
+    constructor(
+        IVault _vault,
+        IProtocolFeeController _newFeeController
+    ) ProtocolFeeControllerMigration(_vault, _newFeeController) SingletonAuthentication(_vault) {
+        _migrationRole = IAuthentication(address(newFeeController)).getActionId(
+            ProtocolFeeController.migratePool.selector
+        );
+
+        // Grant permission required in `migratePools`.
+        _authorizer.grantRole(_migrationRole, address(this));
+    }
+
+    /**
+     * @notice Migrate pools from the old fee controller to the new one.
+     * @dev THis can be called multiple times, if there are too many pools for a single transaction. Note that the
+     * first time this is called, it will migrate the global fee percentages, then proceed with the first set of pools.
+     *
+     * @param pools The set of pools to be migrated in this call
+     */
+    function migratePools(address[] memory pools) external virtual nonReentrant {
+        if (_finalized) {
+            revert AlreadyMigrated();
+        }
+
+        // Migrate the global percentages only once, before the first set of pools.
+        if (_globalPercentagesMigrated == false) {
+            _globalPercentagesMigrated = true;
+
+            _migrateGlobalPercentages();
+        }
+
+        // This more complex migration allows for pool creators and overrides, and uses the new features in the second
+        // deployment of the `ProtocolFeeController`.
+        //
+        // At the end of this process, governance must still withdraw any leftover protocol fees from the old
+        // controller (i.e., that have been collected but not withdrawn). Pool creators likewise would still need to
+        // withdraw any leftover pool creator fees from the old controller.
+        for (uint256 i = 0; i < pools.length; ++i) {
+            // This function is not in the public interface.
+            ProtocolFeeController(address(newFeeController)).migratePool(pools[i]);
+        }
+    }
+
+    function finalizeMigration() external virtual authenticate {
```

Comment:
Hm.

This is a bit tricky. In V2 we just granted admin access and then we just cycled through the stages in the script until the end.

In this case we also need permission to call `finalizeMigration`. I think that bit kinda defeats the point of the script.

What's the worst thing that can happen if we don't migrate some pool (with or without pool creator?). I'm leaning towards reviving the contract from V2 with stages, populating the inputs in the deployment script, and use it like that.

Then, migrate global percentages would be stage one, migrating (all) the hardcoded pools would be stage 2, and then finalize migration the stage 3.

Reply:
Ok; was trying to avoid that, but it makes sense. Will port it over then.

Reply:
Per discussion, looks like we need to keep the current design. The coordinator works for things that are deterministic (e.g., ordered set of immutable tasks, like upgrading a gauge adder). This depends on a dynamic set of pools that would surely change over the course of the governance vote, so it would be out-of-date by the time we could run it.